File: web/js/game.js
var Entity = function(x,y,size,dir,color){
    this.x = x;
    this.y = y;
    this.size = size;
    this.dir = dir;
    this.color = color;
}

var InputHandler = function(){
    this.l = false;
    this.r = false;
}

InputHandler.prototype.left = function(){
    this.l = true;
    this.r = false;
}

InputHandler.prototype.right = function(){
    this.r = true;
    this.l = false;
}

InputHandler.prototype.reset = function(){
    this.l = false;
    this.r = false;
}

var Game = function(){
    this.ticks = 0;
    this.entities = new Array();
    this.player = new Entity(0,0,16,0,0xFF00FF); 
    this.input = new InputHandler();
};
Game.prototype.startUp = function(ctx, x, y){
    //add walls
    for(var i = 0; i < x;i++){
        ctx.fillStyle = "rgb(200,0,0)";
        ctx.fillRect(i,0,16,16);
        ctx.fillRect(i,y-16,16,16);
    }
    for(var i = 0; i < y;i++){
        ctx.fillStyle = "rgb(200,0,0)";
        ctx.fillRect(0,i,16,16);
        ctx.fillRect(x-16,i,16,16);
    }

}
Game.prototype.render = function(ctx){
    for(var i = 0; i < this.entities.length; i++){
        var ent = this.entities[i];
        ctx.fillStyle = ent.color;
        ctx.fillRect(ent.x,ent.y,ent.size,ent.size);
    }
}

File: game/game.go
//The game package provides basic game entites
package game

import "sync"

type Direction int8

const (
	UP Direction = iota
	LEFT
	DOWN
	RIGHT
)

//Basic Entity struct
type Entity struct {
	X     int64
	Y     int64
	Dir   Direction //Direction
	Color int32     //Color of the player
	Died  bool      //flag if dead
	lock  *sync.Mutex
}

//The game struct is a container for all things in the game
type Game struct {
	Entities map[*Entity]bool //Map with every entity in game
	Running  bool             //Flag if game is running
	Map      []bool
	W        int64 //Width of the map
	H        int64 //Height of the map
	Ps       int64 //Player size
}

func (d Direction) clockwise() Direction {
	if d == RIGHT {
		return DOWN
	}

	if d == DOWN {
		return LEFT
	}

	if d == LEFT {
		return UP
	}

	return RIGHT
}

func (d Direction) counterclockwise() Direction {
	if d == RIGHT {
		return UP
	}

	if d == DOWN {
		return RIGHT
	}

	if d == LEFT {
		return DOWN
	}

	return LEFT
}

//only accepts RIGTH or LEFT
func (e *Entity) SetDir(d Direction) {
	e.lock.Lock()
	defer e.lock.Unlock()
	if d == RIGHT {
		e.Dir = e.Dir.clockwise()
	}

	if d == LEFT {
		e.Dir = e.Dir.counterclockwise()
	}

}

func (g *Game) DeleteEntity(e *Entity) {
	delete(g.Entities, e)
}

//If game is finished
func (g *Game) End() bool {
	end := true

	count := 0

	for e := range g.Entities {
		if !e.Died {
			end = false
			count++
		}
	}

	return end || count == (len(g.Entities)-(len(g.Entities)-1))
}

//Update game
func (g *Game) Update() {
	for e, _ := range g.Entities {
		if e.Died {
			continue
		}

		if e.Dir == UP {
			e.Y -= g.Ps
		}

		if e.Dir == RIGHT {
			e.X += g.Ps
		}

		if e.Dir == DOWN {
			e.Y += g.Ps
		}

		if e.Dir == LEFT {
			e.X -= g.Ps
		}

		if g.Collide(e) {
			e.Died = true
			continue
		}

		g.SetPosition(e)
	}
}

//Set the position on the map, not safe!!!
func (g *Game) SetPosition(e *Entity) {
	g.Map[(e.X/g.Ps)+(e.Y/g.Ps)*g.W] = true
}

//Checks if the point of the map is already setted
func (g *Game) Collide(e *Entity) bool {
	if e.X < 0 || e.X >= (g.W*g.Ps) || e.Y < 0 || e.Y >= (g.H*g.Ps) {
		return true
	}

	return g.Map[(e.X/g.Ps)+(e.Y/g.Ps)*g.W]
}

//Reset map and entities, entities will be marked as alive
func (g *Game) Reset() {
	for i := 0; i < len(g.Map); i++ {
		g.Map[i] = false
	}

	for e, _ := range g.Entities {
		e.Died = false
	}

}

//Create a new Entity
func (g *Game) NewEntity(x, y int64, dir Direction, color int32) *Entity {
	e := new(Entity)
	e.X = x
	e.Y = y
	e.Dir = dir
	e.Color = color
	e.Died = false
	e.lock = new(sync.Mutex)

	g.Entities[e] = true
	g.SetPosition(e)
	return e
}

//Create new game with desired map width, height and size of players
func NewGame(width, height int64, psize int64) *Game {
	g := new(Game)
	g.Entities = make(map[*Entity]bool)
	g.Running = false
	g.Map = make([]bool, (width/psize)*(height/psize))
	g.W = width / psize
	g.H = height / psize
	g.Ps = psize
	return g
}
